sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -417,6 +417,11 @@ object NullPropagation extends Rule[LogicalPlan] {
         case left :: Literal(null, _) :: Nil => Literal.create(null, e.dataType)
         case _ => e
       }
+
+      // If the value expression is NULL then transform the In expression to
+      // Literal(null)
+      case In(Literal(null, _), list) => Literal.create(null, BooleanType)
```

Comment:
For someone read this PR in the future:  we don't need to handle `InSet` here, because `InSet` requires `list` expressions are all literals, so if the `value` expression is null literal, the whole `InSet` is foldable and will be optimized into literal null.